File: 14_http_client_go/dns/dns/main_test.go
package main

import (
	"fmt"
	"testing"
)

func Test(t *testing.T) {
	type testCase struct {
		address   string
		expectErr bool
		expected  string
	}

	runCases := []testCase{
		{"https://boot.dev/learn/learn-python", false, "boot.dev"},
		{"https://youtube.com", false, "youtube.com"},
	}

	submitCases := append(runCases, []testCase{
		{"://example.com", true, ""},
	}...)

	testCases := runCases
	if withSubmit {
		testCases = submitCases
	}

	skipped := len(submitCases) - len(testCases)

	passCount := 0
	failCount := 0

	for _, test := range testCases {
		output, err := getDomainNameFromURL(test.address)

		if test.expectErr && (err != nil) != test.expectErr {
			failCount++
			t.Errorf(`---------------------------------
URL:		%v
ExpectErr:	%v
GotErr:		%v
Fail
`, test.address, test.expectErr, err != nil)
		} else if output != test.expected {
			failCount++
			t.Errorf(`---------------------------------
URL:		%v
Expecting:	%v
Actual:		%v
Fail
`, test.address, test.expected, output)
		} else {
			passCount++
			fmt.Printf(`---------------------------------
URL:		%v
Expecting:  %v
Actual:     %v
Pass
`, test.address, test.expected, output)
		}
	}

	fmt.Println("---------------------------------")
	if skipped > 0 {
		fmt.Printf("%d passed, %d failed, %d skipped\n", passCount, failCount, skipped)
	} else {
		fmt.Printf("%d passed, %d failed\n", passCount, failCount)
	}
}

// withSubmit is set at compile time depending
// on which button is used to run the tests
var withSubmit = true

File: 09_dsa/graphs/dijkstra/nearest/main_test.py
from typing import Optional

from main import next_nearest_node

TestCase = tuple[dict[str, int], set[str], Optional[str]]

run_cases: list[TestCase] = [
    (
        {"Minas Tirith": 4, "Isengard": 2, "Gondor": 3, "Mirkwood": 1},
        {"Minas Tirith", "Gondor"},
        "Gondor",
    ),
    (
        {"Minas Tirith": 1, "Isengard": 2, "Gondor": 2, "Mirkwood": 1},
        {"Minas Tirith", "Gondor"},
        "Minas Tirith",
    ),
]

submit_cases: list[TestCase] = run_cases + [
    ({}, set(), None),
    (
        {"Minas Tirith": 1, "Isengard": 2, "Gondor": 2, "Mirkwood": 1},
        {"Isengard", "Mirkwood"},
        "Mirkwood",
    ),
    (
        {
            "Minas Tirith": 3,
            "Isengard": 8,
            "Gondor": 7,
            "Mirkwood": 12,
            "Rivendell": 10,
        },
        {"Isengard", "Mirkwood"},
        "Isengard",
    ),
]


def test(
    distances: dict[str, int], unvisited: set[str], expected: Optional[str]
) -> bool:
    try:
        print("---------------------------------")
        print("Inputs:")
        print(f"- Distances: {distances}")
        print(f"- Unvisited: {unvisited}\n")
        print(f"Expecting: {expected}")
        result = next_nearest_node(distances, unvisited)
        print(f"Actual: {result}\n")
        if result == expected:
            print("Pass")
            return True
        print("Fail")
        return False
    except Exception as e:
        print("Fail")
        print(e)
        return False


def main():
    passed = 0
    failed = 0
    skipped = len(submit_cases) - len(test_cases)
    for test_case in test_cases:
        correct = test(*test_case)
        if correct:
            passed += 1
        else:
            failed += 1
    if failed == 0:
        print("============= PASS ==============")
    else:
        print("============= FAIL ==============")
    if skipped > 0:
        print(f"{passed} passed, {failed} failed, {skipped} skipped")
    else:
        print(f"{passed} passed, {failed} failed")


test_cases = submit_cases
if "__RUN__" in globals():
    test_cases = run_cases

main()
